### DISCUSSION TITLE
How to debug intermittent font/text rendering issue?

### DISCUSSION THREAD

**Discussion by NiklasBr** [gotenberg/gotenberg#861] [Q&A] [answered]
All tips on debugging this issue are very welcome!

Every so often a PDF generated with Gotenberg 8 will lack all fonts loaded with CSS `@font-face { font-family: ...` for the text. It will otherwise have rendered properly, as seen here:

![image](https://github.com/gotenberg/gotenberg/assets/279826/e1168508-970a-4ac5-9638-7422c6fa02f4)

Output is reported as:
```
PDF Producer: Skia/PDF m122
PDF Version: 1.4 
```

We tried [implementing](https://gotenberg.dev/docs/routes#wait-before-rendering-chromium) `waitForExpression` as `'document.readyState === "complete"'`. We tried `window.status` as mentioned in the docs, but it did not change anything (and it is apparently deprecated anyway).

We also tried a dumb regular `waitDelay` but it also had no effect.

Trying to reproduce it on the very same data/request a few minutes later will often lead to a fully working PDF with the fonts and everything 100% as expected:

![image](https://github.com/gotenberg/gotenberg/assets/279826/8891fe99-e079-489d-be06-83a8a667dc79)

**Comment by gulien** [comment]
Hey @NiklasBr 👋 

Could you share the requests for both documents? My first supposition is that you’re using `skipNetworkIdleEvent` 🤔

**Reply by NiklasBr** [reply]
(un?)fortunately we are not using `skipNetworkIdleEvent`. This is the method used, slightly abbreviated:

```php
use Gotenberg\Gotenberg as GotenbergAPI;
use Gotenberg\Stream;

[…]

    protected array $options = [
        'pageWaitFor' => "900ms",
        'displayHeaderFooter' => true,
        'headerTemplate' => null,
        'footerTemplate' => null,
        'format' => 'A4',
        'height' => 11.7,
        'landscape' => false,
        'omitBackground' => false,
        'preferCSSPageSize' => true,
        'printBackground' => true,
        'width' => 8.27,
        'scale' => 1.0,
    ];

    public array $settings = [
        'printBackground',
        'landscape',
        'preferCssPageSize',
        'omitBackground',
        'emulatePrintMediaType',
        'emulateScreenMediaType',
    ];

    public function getOption(string $key): mixed
    {
        return $this->options[$key] ?? null;
    }

    public function getPdfFromHtml(string $renderedHtml, array $params = []): string
    {
        /* Settings from application from Config::getWeb2PrintConfig looks like:
        {
          "displayHeaderFooter": true,
          "format": "A4",
          "landscape": false,
          "marginTop":  0.0,
          "marginBottom":  0.0,
          "marginLeft":  0.0,
          "marginRight": 0.0,
          "omitBackground": false,
          "pageWaitFor": 10,
          "preferCSSPageSize": true,
          "printBackground": false,
          "usePrintCss": true,
          "gotenbergHostUrl": "http://gotenberg:3000"
        }
        */
        $web2PrintGotenbergSettings = Config::getWeb2PrintConfig();
        $gotenbergSettings = json_decode($web2PrintGotenbergSettings['gotenbergSettings'], true);

        $chromium = GotenbergAPI::chromium($web2PrintGotenbergSettings['gotenbergHostUrl'])->pdf();
        $chromium
            ->paperSize(
                $this->getOption('width'),
                $this->getOption('height')
            )
            ->margins(
                $gotenbergSettings['marginTop'],
                $gotenbergSettings['marginBottom'],
                $gotenbergSettings['marginLeft'],
                $gotenbergSettings['marginRight'],
            )
            ->scale(
                $this->getOption('scale')
            )
            ->waitForExpression(
                // https://gotenberg.dev/docs/routes#wait-before-rendering-chromium
                'document.readyState === "complete"'
            );

        foreach ($this->settings as $option) {
            if (isset($params[$option]) && $params[$option] !== false) {
                $chromium->$option();
            }
        }

        if (isset($params['nativePageRanges'])) {
            $chromium->nativePageRanges($params['nativePageRanges']);
        }

        foreach (['header', 'footer'] as $item) {
            if (isset($params[$item . 'Template'])) {
                $chromium->$item(Stream::string(uniqid($item).'.html', $params[$item . 'Template']));
            }
        }

        $request = $chromium->html(Stream::string('processor.html', $renderedHtml));

        return GotenbergAPI::send($request)->getBody()->getContents();
    }
```

**Reply by gulien** [reply]
Thanks! Are your fonts loaded from an external URL (e.g. Google fonts)? Maybe the fonts somehow fail to load on occasion?

**Reply by NiklasBr** [reply]
Fonts, images, styles are all hosted and loaded on the same server, no CDN.

**Reply by gulien** [reply]
No idea what's going on 🤔 Chromium should wait for all assets to be loaded. Do you see anything on the debug logs?
If you use a "standard" font like Arial, does it work every time?

**Reply by NiklasBr** [reply, marked as answer]
It seems standard fonts work, the header and footer are both using `font-family: 'Helvetica Neue', Helvetica, Roboto, Arial, sans-serif;`

I suppose a workaround could be to rebuild the Docker container like so: https://gotenberg.dev/docs/configuration#fonts but unfortunately I only have WOFF/WOFF2 files at the moment and not TTF.

Any detailed logging we could enable?

**Reply by gulien** [reply]
> I suppose a workaround could be to rebuild the Docker container

That would definitely help 👍

You could use `—log-level=debug` to check if yours fonts are correctly loaded in Chromium.

**Reply by NiklasBr** [reply]
Where should I expect logs be stored? Can't find any in `/var/logs` in the running container

**Reply by gulien** [reply]
On the Docker logs directly!

**Reply by NiklasBr** [reply]
@gulien Thanks for your efforts at helping us, docker logs did not provide anything we could discern as clues as to why it happened, but we could resolve the issue by building the Docker images with the TTF font files.

**Reply by gulien** [reply]
@NiklasBr thanks for the feedback!

**Comment by danbuhler** [comment]
We ran into this same issue. Here's how we fixed it (at least sure enough to push into prod)

**`waitForExpression` set to a sync-pollable flag that flips after `document.fonts.ready` resolves, gated on `window.load`** — plus `font-display: swap` as a belt-and-suspenders fallback. The earlier attempts in this thread land slightly too early.

## Symptom

Webfonts loaded via `@font-face` from Google Fonts / Fontshare intermittently didn't render — same input, retried minutes later, worked fine. In our case `font-display: block` made the failure mode dramatic: PDFs came out with a literally empty content stream (864 bytes, `pdffonts` showing no embedded font, zero glyphs drawn). With `display: swap` it would have manifested as silently being the wrong font instead.

## What worked

**Step 1 — inject this script before `</body>`:**

```html
<script>(function(){
  window.__fontsReady = false;
  function wait(){ document.fonts.ready.then(function(){ window.__fontsReady = true; }); }
  if (document.readyState === 'complete') wait();
  else window.addEventListener('load', wait);
})()</script>
```

**Step 2 — on the Gotenberg request, pass:**

```
waitForExpression=window.__fontsReady===true
```

**Step 3 (belt-and-suspenders) — use `font-display: swap`, not `block`:**

```html
<link href="https://fonts.googleapis.com/css2?family=Lora&display=swap" rel="stylesheet">
```

With the wait in place, `swap` is strictly safer than `block`. If `waitForExpression` ever does time out (Gotenberg's default `API_TIMEOUT` is 15s), `block` would emit a PDF with invisible text — `swap` falls back to a system font instead, which is at least readable. The wait is the primary mechanism; `swap` is the safety net for the edge case where it ever fails.

## Empirical results

Forcing the race with 10 different Google Fonts in a row against a freshly restarted Gotenberg (so each font is cold-cache):

- **Before the fix: 10/10 PDFs had zero glyphs** (864 bytes each).
- **After the fix: 0/10 broken**, all PDFs include the proper font subset (13–46 KB).

Cost: ~100–300 ms per render on a cold font, which is exactly the woff2 download time — i.e. we're correctly waiting for the binary instead of capturing prematurely.

## Other notes

- Tested on Gotenberg 8.27.0 and 8.29.1 with identical results — the bug is in Chromium's networkIdle-vs-FontFaceSet timing, not Gotenberg itself.
- The race is *per-tab*: Gotenberg's default `--chromium-max-concurrency=6` means each conversion lands in a renderer process whose tab needs to register the font on its own, even when the woff2 bytes are already in the shared HTTP cache. This is why the issue can keep firing throughout a session, not just on cold start — and why other workarounds based on "cache the fonts first" don't fully solve it.
- The "bundle TTFs in the container" workaround recommended elsewhere in this thread also works, but only if your font list is fixed at build time. We have a user-driven font picker (~1,800 Google Fonts), so pre-baking wasn't an option — this `waitForExpression` approach is essential when fonts are dynamic.

Hope this helps anyone else hitting the same thing.

**Reply by gulien** [reply]
@danbuhler thanks! Just to be sure to understand, setting skipNetworkIdleEvent to `false` does not fix the issue?

**Reply by danbuhler** [reply]
@gulien thanks for asking. I didn't try that. I thought the default was false. (incorrectly on v8 it appears)

**Reply by gulien** [reply]
The default was indeed `false`, but for most users the `true` value is actually better (faster conversion).  I changed the default a while ago.

See https://gotenberg.dev/docs/convert-with-chromium/convert-url-to-pdf#network-errors
